### PR TITLE
Fix possibility of archive entry being extracted to outside extract destination

### PR DESCRIPTION
**Pull request by TranceLove**
Details about the vulnerability: [here](https://snyk.io/research/zip-slip-vulnerability); fix for 3.2.1 is #1260 which had been released to Google Play Store as version 3.2.2.

Also brought the test suite from e0d40191e in #1222 for behavioural verification.

IOException will be thrown on erroneous archive entry, causing app to pop up an "Error" Toast.

Safe to cherry-pick into `release-3.3.0`.

Our sincere thanks to folks at [snyk.io](https://snyk.io) who tipped us about the vulnerability.

### CODE REVIEW

**Comment by TranceLove** [comment]
For fix against 3.2.1 the current Google Play Store release, see [TranceLove:AmazeFileManager/bugfix/fix-b0rkenzip-321](https://github.com/TranceLove/AmazeFileManager/tree/bugfix/fix-b0rkenzip-321).

Since 3.2.1 codebase doesn't in favour of writing unit tests, verification was done against Pixel 2 emulators running 5.1.1, 6.0, 7.1.1, 8.1, plus Oneplus One running Slim7 (7.1.2), Galaxy S2 running SlimLP (5.1.1) and GPD XD running LegacyROM (4.4.4).

**Comment by EmmanuelMess** [comment]
@TranceLove PR [TranceLove:fix-b0rkenzip-321](https://github.com/TranceLove/AmazeFileManager/tree/bugfix/fix-b0rkenzip-321) into [hotfix-zip](https://github.com/TeamAmaze/AmazeFileManager/tree/hotfix-zip).

File: app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java [hunk 1]
```diff
@@ -28,6 +28,7 @@
      * e.g. rar internally uses '\' but is converted to "/" for the app.
      */
     public static final String SEPARATOR = "/";
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/compressed/CompressedHelper.java]: This should be `SEPARATOR = SEPARATOR_CHAR + "";`.

**Comment by EmmanuelMess** [comment]
In the win file, it seems to try to unzip the bad file (try to open it and then try to open anything else, check the notifications).

**Comment by TranceLove** [comment]
Added 9722b0d which ports 9888ae7 from #1260 for `hotfix-zip`.

Unit test also updated to reflect change (team members only).

**Comment by TranceLove** [comment]
Added 3a88087. Unit test also updated to reflect changes (team members only - it will appear in PR once it's approved)

**Comment by TranceLove** [comment]
If you need a Dialog, you'll need to find a way to display a dialog in ExtractService.DoWork, best at `onInvalidEntriesFoundBeforeStart(invalidArchiveEntries)`...

**Comment by TranceLove** [comment]
See if it's what you'd expect in 045a51e.

**Comment by VishalNehra** [comment]
@EmmanuelMess

**Comment by EmmanuelMess** [comment]
@VishalNehra

**Comment by EmmanuelMess** [comment]
We merge this and release 3.3.0, @VishalNehra please review.

**Comment by VishalNehra** [comment]
I started a review for this. Also, we should wait for #1285 to pass before releasing v3.3.0

**Comment by VishalNehra** [comment]
@TranceLove please merge this with master and see the review I started.

**Comment by TranceLove** [comment]
@VishalNehra sorry but seems I couldn't find a review here... would it disappear in a force update?

However, if you're talking about this line...
https://github.com/TranceLove/AmazeFileManager/blob/3f025bf5f59dd301c845adf4930d0bc1eb5ecf14/app/src/main/res/values/strings.xml#L643

Please see if you need a better wording for this.

**Comment by VishalNehra** [comment]
My bad, might have been network problem.

**Comment by VishalNehra** [comment]
Can you still not see the review? Basically I asked to use ````replaceAll```` instead of ````replace```` in ````Extractor#fixEntryName````

**Comment by TranceLove** [comment]
Sorry but really couldn't see it... but is done at dac7948.

Also added B0rkenZipTest since PR is closing in soon.

**Pull request merged**